src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java

```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
 	
 	// TEST
 	
-	public void main(String[] args) {
+	public static void main(String[] args) {
 		Log.start();
 		FirmwareUploader fu = new FirmwareUploader();
 		try {
-			fu.run("./firmware.hex", "COM3");
+			fu.run("./firmware.hex", "COM3", new SelectTextArea("test","test","") );
 		} catch(Exception e) {
 			e.printStackTrace();
 		}
 	}
+	
+	//
+	//
+	//
+	
+	public static SimpleDateFormat simpleDateFormat = new SimpleDateFormat("ddMMyyyy hh:mm:ss");
+
+	static boolean debugRuntimeExecPre = true;
+	static boolean lastExecSucces = false;
+
+	// TODO ! a return 
+	class ExecBashCommandResult{
+	    boolean haveStart = true;
+	    boolean isFinis = false;
+	    Integer returnCode = null;
+	    String logs = "";
+	    
+	    //TODO 
+	    
+	    
+	}
+	// TO REVIEW / TODO le cas d'un process interactif ou sans fin ...
+	// TO REVEIW ( as this is a way to exec commands on the user system ... )
+	// should be a least private for security (and all methodes that using it ) ?
+	protected static String execBashCommand(String[] cmdArray, StreamGobblerReadLineBufferedSpecial streamGobblerProcessIn,SelectTextArea  execResult) {
+	    lastExecSucces = false;
+	    final StringBuffer res = new StringBuffer();
+	    res.setLength(0);
+	    if (debugRuntimeExecPre) {
+		System.out.printf("Running : ");
+		for (String arg : cmdArray) {
+		    System.out.printf("%s ", arg);
+		}
+		System.out.printf("\n");
+	    }
+	    res.append("Running : ");
+	    for (String arg : cmdArray) {
+		res.append(String.format("%s ", arg));
+	    }
+	    res.append("\n");
+
+	    try {
+		Date dStar = new Date();
+		Process process;
+		ProcessBuilder pb = new ProcessBuilder(cmdArray);
+		//Map<String, String> env = pb.environment();
+		//env.put("VAR1", "myValue");
+		//env.remove("OTHERVAR");
+		//env.put("VAR2", env.get("VAR1") + "suffix");
+		//pb.directory(new File("myDir"));
+
+		// To merge err on out // TO REVIEW for specials case.
+		pb = pb.redirectErrorStream(true);
+		
+		//Process 
+		process = pb.start();
+
+		if (streamGobblerProcessIn == null) {
+		    streamGobblerProcessIn = new StreamGobblerReadLineBufferedSpecial(process.getInputStream(), "out") {
```

Comment:
I think the flash process should not ask anything to the user thus capturing the inputstream seems not necessary, what do you think ?